### PR TITLE
Add stacks blockchain support

### PR DESCRIPTION
**Pull request by emmanuelm41**
### 📝 Description

This PR intents to add stacks blockchain support for sending, receiving and checking balance features.

### ❓ Context

- **Impacted projects**: `common, cli, desktop, mobile, cryptoassets` 
- **Linked resource(s)**: `` 

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/ledger-live-common/src/families/stacks/utils.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { ResponseAddress } from "@zondax/ledger-stacks";
+import { isValidBase64, isValidHex } from "../filecoin/utils";
+
+export const isNoErrorReturnCode = (code: number) => code === 0x9000;
+
+export const getPath = (path: string) =>
+  path && path.substr(0, 2) !== "m/" ? `m/${path}` : path;
+
+export const isError = (r: ResponseAddress) => {
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/utils.ts:9]: Done!

File: libs/ledger-live-common/src/families/stacks/bridge/account.ts [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+          if (!nonce) {
+            throw new InvalidNonce();
+          }
+
+          const blockstack = new BlockstackApp(transport);
+
+          o.next({
+            type: "device-signature-requested",
+          });
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/account.ts]: Done!

File: libs/ledger-live-common/src/families/stacks/bridge/utils/api.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    newOffset = 0;
+  let txs: TransactionResponse[] = [];
+
+  while (!done) {
+    const { results, total, limit } = await fetchTxs(addr, newOffset);
+    txs = txs.concat(results);
+    newOffset += limit;
+
+    if (newOffset >= total) done = true;
+  }
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/utils/api.ts:119]: nice change

File: libs/ledger-live-common/src/families/stacks/hw-signMessage.ts
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/hw-signMessage.ts:1]: This file is used to run the signMessage cmd on the CLI

File: libs/ledger-live-common/src/families/stacks/specs.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import type { AppSpec } from "../../bot/types";
+import { acceptTransaction } from "./speculos-deviceActions";
+
+const MIN_SAFE = new BigNumber(10000);
+const stacksSpecs: AppSpec<Transaction> = {
+  disabled: false,
+  name: "Stacks",
+  currency: getCryptoCurrencyById("stacks"),
+  appQuery: {
+    model: DeviceModelId.nanoS,
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/specs.ts:18]: Done!

File: libs/ledger-live-common/src/families/stacks/types.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+export type TransactionStatusRaw = TransactionStatusCommonRaw;
+
+export type BroadcastFnSignature = (arg0: BroadcastArg0) => Promise<Operation>;
+
+export type CoreStatics = Record<string, never>;
+export type CoreAccountSpecifics = Record<string, never>;
+export type CoreOperationSpecifics = Record<string, never>;
+export type CoreCurrencySpecifics = Record<string, never>;
+export const reflect = (_declare: any) => {};
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/types.ts:54]: Removed!

File: apps/ledger-live-mobile/src/components/RootNavigator/types/SendFundsNavigator.ts [hunk 6]
```diff
@@ -335,4 +336,36 @@ export type SendFundsNavigatorStackParamList = {
...
+    currentNavigation:
+      | ScreenName.SignTransactionSummary
+      | ScreenName.SendSummary
+      | ScreenName.SwapForm;
+    nextNavigation:
+      | ScreenName.SignTransactionSelectDevice
+      | ScreenName.SendSelectDevice
+      | ScreenName.SwapForm;
+  };
+  [ScreenName.StacksEditMemo]: {
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply apps/ledger-live-mobile/src/components/RootNavigator/types/SendFundsNavigator.ts:354]: Yes, right!

File: apps/ledger-live-mobile/src/families/stacks/SendRowsCustom.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import type { Account } from "@ledgerhq/types-live";
+import { Transaction } from "@ledgerhq/live-common/generated/types";
+import SendRowMemo from "./SendRowMemo";
+import {
+  BaseComposite,
+  StackNavigatorProps,
+} from "../../components/RootNavigator/types/helpers";
+import { LendingEnableFlowParamsList } from "../../components/RootNavigator/types/LendingEnableFlowNavigator";
+import { LendingSupplyFlowNavigatorParamList } from "../../components/RootNavigator/types/LendingSupplyFlowNavigator";
+import { LendingWithdrawFlowNavigatorParamList } from "../../components/RootNavigator/types/LendingWithdrawFlowNavigator";
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply apps/ledger-live-mobile/src/families/stacks/SendRowsCustom.tsx:11]: Done!

File: libs/ledger-live-common/src/families/stacks/bridge/account.ts [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+  return new BigNumber(balance.balance);
+};
+
+const prepareTransaction = async (
+  a: Account,
+  t: Transaction
+): Promise<Transaction> => {
+  const { id: accountID, balance } = a;
+  const { recipient, useAllAmount } = t;
+  const { xpubOrAddress: xpub } = decodeAccountId(accountID);
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/account.ts:140]: For some reason I am not sure now, this was not enough to find the xpub value. That is why I took this way.

File: libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts:1]: From my point of view, `utils` inside bridge folder is really specific to that part of the code. It is not being used by functions outside that folder. While in the `utils.ts` in stacks folder, I have really generic functions.

File: libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore need to ignore the TS error here
+  tx.auth.spendingCondition.signature = createMessageSignature(signature);
+
+  return tx.serialize();
+};
+
+export const getUnit = () => getCryptoCurrencyById("stacks").units[0];
+
+export const getAddress = (a: Account): Address =>
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts:58]: change the name for something generic like `misc.ts`

File: libs/ledger-live-common/src/families/stacks/bridge/utils/types.ts
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/bridge/utils/types.ts:1]: Yes, I like this!

File: libs/ledgerjs/packages/errors/src/index.ts [hunk 10]
```diff
@@ -76,8 +76,10 @@ export const FirmwareNotRecognized = createCustomErrorClass(
 export const HardResetFail = createCustomErrorClass("HardResetFail");
 export const InvalidXRPTag = createCustomErrorClass("InvalidXRPTag");
 export const InvalidAddress = createCustomErrorClass("InvalidAddress");
-export const InvalidAddressBecauseDestinationIsAlsoSource =
-  createCustomErrorClass("InvalidAddressBecauseDestinationIsAlsoSource");
+export const InvalidNonce = createCustomErrorClass("InvalidNonce");
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledgerjs/packages/errors/src/index.ts:79]: That is perfect! I added the description on English for that error on LLD and LLM. I think the error is generic enough to be useful for other chains

File: apps/ledger-live-desktop/src/renderer/families/stacks/MemoValueField.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  const onMemoValueChange = useCallback(
+    memoValue => {
+      onChange(bridge.updateTransaction(transaction, { memo: memoValue }));
+    },
+    [onChange, transaction, bridge],
+  );
+
+  // We use transaction as an error here.
+  // It will be usefull to block a memo wrong format
+  // on the ledger-live mobile
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply apps/ledger-live-desktop/src/renderer/families/stacks/MemoValueField.tsx:36]: Thanks! Good catch!

File: libs/ledger-live-common/src/families/stacks/specs.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+          expect(account.balance.toFixed()).toBe(
+            accountBeforeTransaction.balance.minus(operation.value).toFixed()
+          )
+        );
+      },
+    },
+    {
+      name: "Transfer Max",
+      maxRun: 1,
+      transaction: ({ account, siblings, bridge }) => {
```
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/specs.ts:65]: Yes, let me do it!

File: libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledger-live-common/src/families/stacks/bridge/utils/utils.ts:1]: Makes sense of course, it's just the same practical issue as a dev than with all `index.ts` files, when you're trying to look for a specific one but there are so many of them :slightly_smiling_face: 
No big deal let's keep it that way.

File: libs/ledger-live-common/src/families/stacks/hw-signMessage.ts
[the remark this answers is not in the document]
Comment (haammar-ledger) [reply libs/ledger-live-common/src/families/stacks/hw-signMessage.ts:1]: Noted, the file doesn't hurt no problem, but for my own curiosity, what's the use-case for the signMessage command? it's not mandatory from Live perspective, is there any need to use it from CLI?

[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/hw-signMessage.ts:1]: Hakim, check it [here](https://github.com/LedgerHQ/ledger-live/blob/235de21681ddb3447fd800a0309eb75ad763ea00/apps/cli/src/commands/signMessage.ts#L63). The `signMessage` CLI command is using it.

File: libs/ledger-live-common/src/families/stacks/specs.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (emmanuelm41) [reply libs/ledger-live-common/src/families/stacks/specs.ts:65]: @haammar-ledger what do you mean by a test case? I thought you wanted me to add another test case on the specs (like 50% transfer), but I see we have two there.

**Comment by haammar-ledger** [comment]
Replaced with https://github.com/LedgerHQ/ledger-live/pull/3459

**Pull request closed without merging**